DANS-KNAW/easy-deposit-api
# EASY-1657 post file using content-disposition header

Fixes EASY-1657 Implement upload single file with POST directory

#### When applied it will
* interpret the specs as sub-requests of a request with `multipart/form-data`
* `Content-Type` of the sub-requests are still ignored, so zip-files [EASY-1658](https://drivenbydata.atlassian.net/browse/EASY-1658) are uploaded as-is.
* The response code on success is 200-ok as it might be a mix of created and replaced. The specs says 201-created but that applies to subrequests. No specification how to wrap these up in the `multipart/form-data` request.

Some statistics:

#### Where should the reviewer @DANS-KNAW/easy start?

#### How should this be manually tested?

* see the updated `test/resources/readme.md`
* test both PUT and POST
* test with curl: see this [stack-overflow](https://stackoverflow.com/questions/4238809/example-of-multipart-form-data)
* [x] large files https://github.com/DANS-KNAW/easy-deposit-api/pull/63#issuecomment-413111374

Uploading a file of 540 MB to `deasy` resulted in the browser complaining about "the connection was reset" and the log below (`HttpParser full`, replaced IP addresses with ...). After clearing the `bag/data` folder on the server I could upload it multiple times. Same for 1GB and even the 5GB succeeded on a second attempt. Might have to to with fooling around with the browser's back-button on the test page while changing the file to upload. It looks like a proper reset is not only clear `bag/data` but also POST a small file.

```
[2018-08-23 15:43:42,638] WARN  HttpParser Full for SCEP@747f9023{l(/...:59536)<->r(/...:20190),s=1,open=true,ishut=false,oshut=false,rb=false,wb=false,w=true,i=0r}-{AsyncHttpConnection@67754c1c,g=HttpGenerator{s=0,h=-1,b=-1,c=-1},p=HttpParser{s=-13,l=0,c=-3},r=1}
```
Some statistics of a successful upload of 5GB. Of course statistics with `teasy` would be more realistic that with `deasy`.

```
[2018-08-24 09:10:43,910] INFO  uploading to [/var/opt/dans.knaw.nl/deposit/drafts/user001/196fb09a-03c5-4951-a308-4879e4b7f2ab/bag] of [path/tox/large.txt]
[2018-08-24 09:12:37,632] INFO  created=true user001/196fb09a-03c5-4951-a308-4879e4b7f2ab/path/tox/large.txt
[2018-08-24 09:12:37,633] INFO  POST returned status=200 headers=Map()
-rw-r--r--. 1 easy-deposit-api easy-deposit-api 1663       2018-08-24 09:12:37.624723161 +0200 manifest-sha1.txt
-rw-r--r--. 1 easy-deposit-api easy-deposit-api 5007600000 2018-08-24 09:11:50.768755240 +0200 large.txt

upload:   10:43 tot 11:50 = 1:07
manifest: 11:50 tot 12:37 = 0:47
```

#### Related pull requests on github
~~May cause merge conflicts on IntegrationSpec.java with #62~~ merged
